### PR TITLE
Wrap the test func with try/except to always call destroy_process_group

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129514
* #129512
* #129511
* __->__ #124961

This can avoid PG warning about not calling destry_pg

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]
Cool! this would solve the recent poping warning if test crashes

**Review by wz337 — APPROVED** [review]

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 1]
```diff
@@ -383,8 +383,12 @@ def wrapper(
...
 
         self.init_pg()
-        func(self, *args, **kwargs)  # type: ignore[misc]
-        self.destroy_pg()
+        try:
+            func(self, *args, **kwargs)  # type: ignore[misc]
+        except Exception as e:
+            raise e
+        finally:
+            self.destroy_pg()
```
Comment (awgu) [inline torch/testing/_internal/distributed/_tensor/common_dtensor.py:391]: It looks like `self.destroy_pg()` is:
```
dist.barrier()
dist.destroy_process_group()
```

IIUC, `func()` is being run in the calling process and does not spawn any new process or thread. What happens if one rank raises an exception but another rank does not? Would the `barrier()` hang in `self.destroy_pg()`?

(For context, I want to get rid of this warning for FSDP tests too, but I was not sure if this approach will always work.)

Comment (fegin) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py]: Ideally, we should not have to call `barrier()` before calling `destroy_pg()`, but I'm not sure if that is the case. cc., @wconstab . However, this version is still better than the original one. If a rank raises while other don't, the original code would have the same issue that one rank doesn't participate in the `barrier()` call.

**Pull request closed without merging**